### ISSUE TITLE
Flattening of extends with self-references

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1458]
**Reported by perost on 9 Apr 2014 13:19 UTC**
Consider the following model:
```
model T1
  Real x;
end T1;

model T2
  Real x = 0.0;
  Real y = 0.0;
end T2;

model A
  replaceable T1 t;
end A;

model B
  extends A(redeclare T2 t(x = t.y));
end B;
```
The spec says that when flattening B we should look up the base class A, flatten it and insert it into the class. To be able to flatten A we need to apply the modifier on it. The modifier references t.y though, and to look up that name we need to flatten A. Does this case break the rule that it should be possible to flatten all extends-clauses before inserting the result of their flattening, or is there some way of handling this?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1458

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 9 Apr 2014 13:33 UTC**
Replying to [ticket:1458 perost]:
> The spec says that when flattening B we should look up the base class A, flatten it and insert it into the class. To be able to flatten A we need to apply the modifier on it. 

Technically it says that we need to merge the modifiers to A - not "apply it", i.e. it does not state that we need to evaluate the modifier yet; and thus we can delay it.

In this case the merging is kind of obvious - since the declaration does not have any.

> The modifier references t.y though, and to look up that name we need to flatten A. Does this case break the rule that it should be possible to flatten all extends-clauses before inserting the result of their flattening, or is there some way of handling this?

It shouldn't break that rule - the rule is intended to guard against lookup being messed up, e.g. by having model A with an A inside it.

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 11 Apr 2014 05:03 UTC**
> Technically it says that we need to merge the modifiers to A - not "apply it"

Well, the specification text is:

5.6.1 Partial Flattening
First the names of declared local classes and components are found. Here modifiers are merged to the local elements and redeclarations take effect (including redeclared elements, see Section 7.3). 
Then base-classes are looked up, flattened and inserted into the class ...

5.6.2 Flattening
Partially flatten the class, apply the modifiers (Section 7.2) and flatten all local elements.

So 5.6.1 say that base-classes are looked up, flattened and 5.6.2 
say that one has to partially flatten the class, apply the modifiers.

This is a bit wrong as partially flatten the class contains merging of modifiers.
What does "apply the modifers" in 5.6.2 means then?

Basically what you're saying is that modifiers are merged blindly without even checking that they exist (no lookup) until the flattening is done.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 11 Apr 2014 07:17 UTC**
Replying to [comment:2 adrpo]:
> > Technically it says that we need to merge the modifiers to A - not "apply it"
> 
> Well, the specification text is:
> 
> 5.6.1 Partial Flattening
> First the names of declared local classes and components are found. Here modifiers are merged to the local elements and redeclarations take effect (including redeclared elements, see Section 7.3). 
> Then base-classes are looked up, flattened and inserted into the class ...
> 
> 5.6.2 Flattening
> Partially flatten the class, apply the modifiers (Section 7.2) and flatten all local elements.
> 
> So 5.6.1 say that base-classes are looked up, flattened and 5.6.2 
> say that one has to partially flatten the class, apply the modifiers.
> 
> This is a bit wrong as partially flatten the class contains merging of modifiers.
> What does "apply the modifers" in 5.6.2 means then?

Well, merging is well-defined, so apply is something else and I agree that it needs to be better defined.

Since redeclarations are done earlier the remaining part is that apply means setting the values, i.e. for "x=5" when flattening "x" we apply the modifier to bind x to 5.

It would make more sense if redeclare would take effect in this step as well.

> Basically what you're saying is that modifiers are merged blindly without even checking that they exist (no lookup) until the flattening is done. 

Well, that would be one way of seeing it. Another is that some parts can be checked earlier - but some parts require that you flatten the model first.

Specifically:
There can be errors during merging - they could be reported then, or collected and reported later.
The modifier may modify something that cannot be modified - that can be detected either during flattening or afterwards.
But missing elements must be done after flattening, yes.

If you have 
```
 model M
   A a(x=a.y+b.y.x);
   B b(x=a.x+b.y.x, redeclare Y2 y);
 end M;
 M m(a.x.start=2);
```
You cannot know that A and B contain x and y until you have flattened them - and before that you cannot know that the references in the modifiers actually exists. 
You can still merge the new modifier.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 11 Apr 2014 13:03 UTC**
Ok, so consider this example instead:
```
model T1
  Real x = 1.0;
end T1;

model M
  model T2
    Real x = 2.0;
  end T2;
end M;

model A
  replaceable T1 t;
end A;

model B
  extends A(redeclare T2 t);
  extends M;
end B;
```
In this example, when should the redeclare modifier on A be applied? I assume that "T2 t" must be flattened in the scope of B, but that can't be done before flattening the extends since T2 is inherited from M. And to flatten the baseclass A it seems like we need to flatten the redeclaration first, since A should be flattened before inserting its contents into B. And indeed, Dymola says that T2 couldn't be found when I tried it. So is this example somehow illegal?

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 11 Apr 2014 15:33 UTC**
Replying to [comment:4 perost]:
> Ok, so consider this example instead:
> ```
> model T1
>   Real x = 1.0;
> end T1;
> 
> model M
>   model T2
>     Real x = 2.0;
>   end T2;
> end M;
> 
> model A
>   replaceable T1 t;
> end A;
> 
> model B
>   extends A(redeclare T2 t);
>   extends M;
> end B;
> ```
> In this example, when should the redeclare modifier on A be applied? 

The redeclare should take effect during partial flattening.

Thus we cannot even partially flatten the base-classes independently of each other and in any order and get the same result; so the example is illegal.
(Yes, there are some issues to improve in that chapter to clear up semantics without trying to change it. See #1397)

>I assume that "T2 t" must be flattened in the scope of B, but that can't be done before flattening the extends since T2 is inherited from M. And to flatten the baseclass A it seems like we need to flatten the redeclaration first, since A should be flattened before inserting its contents into B. And indeed, Dymola says that T2 couldn't be found when I tried it. So is this example somehow illegal?

Yes, currently.

However, it would be possible to specify that the redeclare is done later - I believe it would be cleaner and more general.

A current work-around is to use:
```
model B
  extends A;
  extends B;
  redeclare T2 t;
end B;
```
The disadvantage is that you cannot know where `t` is declared, and more importantly if there is an error and `t` is is missing you have no idea where it is missing from.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 11 Apr 2014 15:49 UTC**
Replying to [comment:5 hansolsson]:
> A current work-around is to use:
> ```
> model B
>   extends A;
>   extends B;
>   redeclare T2 t;
> end B;
> ```

Not sure I follow correctly here. Did you mean this instead?
```mo
model B
  extends A;
  extends M;
  redeclare T2 t;
end B;
```

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 11 Apr 2014 15:51 UTC**
Replying to [comment:6 jmattsson]:
> Replying to [comment:5 hansolsson]:
> > A current work-around is to use:
> > ```
> > model B
> >   extends A;
> >   extends B;
> >   redeclare T2 t;
> > end B;
> > }}}
> Not sure I follow correctly here. Did you mean this instead?
> {{{#!mo
> model B
>   extends A;
>   extends M;
>   redeclare T2 t;
> end B;
> ```

Yes. Sorry - should have copy-pasted the working example instead of typing it again.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 11 Apr 2014 16:42 UTC**
Replying to [comment:5 hansolsson]:
> A current work-around is to use:
> ```
> model B
>   extends A;
>   extends M;
>   redeclare T2 t;
> end B;
> ```

Shouldn't that redeclare take effect at the same time as the redeclare in my example though? Or do you mean that redeclared elements should be applied after flattening the base classes?

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 12 Apr 2014 08:14 UTC**
Replying to [comment:3 hansolsson]:
> Replying to [comment:2 adrpo]:
> 
> > Basically what you're saying is that modifiers are merged blindly without even checking that they exist (no lookup) until the flattening is done. 
> 
> Well, that would be one way of seeing it. Another is that some parts can be checked earlier - but some parts require that you flatten the model first.
> 
> Specifically:
> There can be errors during merging - they could be reported then, or collected and reported later.
> The modifier may modify something that cannot be modified - that can be detected either during flattening or afterwards.
> But missing elements must be done after flattening, yes.
> 

Or some parts are not checked at all (blindly merged/applied and discarded it seems).
You can try to flatten this model with Dymola and see how many errors you get:

```
model SelfReference1
  package P
    type T = Real(start = 1);
    package L
      replaceable package M = 
        P(T(start = 3, 
            somethingThatIsNotThere = 20, 
            SomethingElseThatIsNotThere(x = z)));

      model X
        parameter M.T t = 10;
      end X;
    end L;
  end P;

  P.L.X x;
end SelfReference1;
```

I would expect this model to give errors that:
- z is not present in P.L
- SomethingElseThatIsNotThere is not found in P.T
- somethingThatIsNotThere is not found in P.T

This is the danger of not checking earlier (where the modifiers are defined)
that their target or their value actually exists. I understand that this might
be complicated because of the way Modelica is defined but I still think it should
be possible to do.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 14 Apr 2014 10:22 UTC**
Here's another interesting model with regards to flattening:
```mo
model A
  replaceable model B
    model C
      Real x;
    end C;
  end B;

  B b;
end A;

model M
  extends A(redeclare model B = B.C);
end M;
```
It doesn't work in any tool I tested, so it's probably not a valid model. Still, I'm curious how this model is supposed to be flattened, even if it leads to an error. More specifically, at what point in the flattening is B.C looked up.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 14 Apr 2014 19:34 UTC**
Replying to [comment:5 hansolsson]:
> Replying to [comment:4 perost]:
> > Ok, so consider this example instead:
> > {{{
> > model T1
> >   Real x = 1.0;
> > end T1;
> > 
> > model M
> >   model T2
> >     Real x = 2.0;
> >   end T2;
> > end M;
> > 
> > model A
> >   replaceable T1 t;
> > end A;
> > 
> > model B
> >   extends A(redeclare T2 t);
> >   extends M;
> > end B;
> > }}}
> > In this example, when should the redeclare modifier on A be applied? 
> 
> The redeclare should take effect during partial flattening.
> 
> Thus we cannot even partially flatten the base-classes independently of each other and in any order and get the same result; so the example is illegal.

Unfortunately, there is no requirement in the Spec right now that says that we should be able to partially flatten the base-classes independently. There is a requirement (in 5.6.1) that the **lookup** of the base-classes should be independent of the order in which they are handled. The lookup seems to refer to looking up the name of the base class, and not the partial flattening. So there is nothing that says that this example is illegal.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 9 May 2014 08:12 UTC**
Replying to [comment:11 eshmoylova]:
> Unfortunately, there is no requirement in the Spec right now that says that we should be able to partially flatten the base-classes independently. There is a requirement (in 5.6.1) that the **lookup** of the base-classes should be independent of the order in which they are handled. The lookup seems to refer to looking up the name of the base class, and not the partial flattening. So there is nothing that says that this example is illegal.

I see it as a bit more complicated: The restriction is in the text but unfortunately part of the not very clear non-normative text in 5.6.1.

We can also make the example worse:
```
model M
  model M2
    Real x=2.0;
  end M2;
  replaceable T1 m;
end M;

model A
  model A3
    Real x=3.0;
  end A3;
  replaceable T1 a;
end A;

model B
  extends A(redeclare M2 a);
  extends M(redeclare A3 m);
end B;
```
The specification is not clear if we handle "extends A" or "extends M" first (the non-normative text states that it should not matter), but it is clear that we do one and then the other (since we first flatten and then insert the result - and before we insert the result lookup will not find anything) - unfortunately that is not possible in this example if we want the desired result.

There are two ways of handling this:
1. Add normative text corresponding to the comment about order not mattering; making this example illegal.
2. Split the description into more phases, such that the lookup of this modifier is done after adding the contents of the extends-clauses, something along the lines of:
5.6.1.1: First the names of declared local classes and components are found. Here modifiers are merged to the local elements. Then base-classes are looked up, (5.6.1.1 is applied to them) and inserted into the class. The lookup of the base-classes should be...
5.6.1.2. Redeclarations take effect (including redeclared elements, see Section 7.3). 
This also needs to be combined with the clarification of #1397.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 16 Jun 2014 03:08 UTC**
I would be in favor of the option 2. It will avoid situations when someone has a working model then while editing the code changes the order of the extends-clauses, and gets a broken model. If it gives an error it would be at least an easy thing to spot. But if, for example, there were top-level definitions of A3 and M2 in the example in the previous comment, there would be no error raised. Yet, if the top-level ones were different from the ones inside A and M changing the order of extends clauses would produce different results.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 16 Jun 2014 11:36 UTC**
Replying to [comment:13 eshmoylova]:
> I would be in favor of the option 2. 

I agree. Also, that rule will be easier to understand, especially for users.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 17 Jun 2014 17:05 UTC**
Spec group:
Jesper: Breadth-first search of base-classes.
Hans:
For
```
model M
  model M2
    Real x=2.0;
  end M2;
  replaceable T1 m;
end M;

model A
  model A3
    Real x=3.0;
  end A3;
  replaceable T1 a;
end A;

model B
  extends A(redeclare M2 a);
  extends M(redeclare A3 m);
end B;
```
Step 1 looks up base-classes A and M
Step 2 finds that A contain model A3 and component a and merge modifier to a;
finds that M contain model M2 and component m and merge modifier to m
Step 3 combines contents of A and M into B

_We can now perform lookup in B_

Step 4 (in any order except for dependencies, and on demand) recurses for elements of B so that they are flattened which for components mean they get a type -  taking redeclare into account; (modifiers will use their original scope see #1397)

Martin (S): Duplicate elements is not specified here.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 17 Jun 2014 17:07 UTC**
Spec group:

Some are unsure about adding this to 3.3r1 and/or the implications.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 18 Jun 2014 09:32 UTC**
Turning the above into spec text:

5.6.1 Partial Flattening
First the names of declared local classes and components are found. Here modifiers are merged to the local elements. Then base-classes are looked up, partially flattened (in the same way) and inserted into the class.

After the partial flattening it is possible to perform lookup in the class, since the name of all local and inherited elements are known. 
5.6.2. Flattening
Partially flatten all elements of the class. Here classes for components are looked up (taking redeclarations into account).

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 18 Jun 2014 10:02 UTC**
Replying to [comment:17 hansolsson]:
> Turning the above into spec text:
> ... 

Maybe I'm missing something, but doesn't this say that inherited elements should be flattened in the local scope? I.e. base-classes are partially flattened and inserted into the class, and then both local and inherited elements are partially flattened in the flattening step. Is that really the intent?

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 18 Jun 2014 10:47 UTC**
Revised 5.6.2 based on discussion at meeting:
All elements are flattened, starting by partially flattening them.

As part of this classes for component elements are looked up, i.e. modifiers including redeclarations for component take effect – and the flattened components are inserted into the resulting class.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 18 Jun 2014 10:49 UTC**
Replying to [comment:18 perost]:
> Replying to [comment:17 hansolsson]:
> > Turning the above into spec text:
> > ... 
> Maybe I'm missing something, but doesn't this say that inherited elements should be flattened in the local scope? I.e. base-classes are partially flattened and inserted into the class, and then both local and inherited elements are partially flattened in the flattening step. Is that really the intent?

This is related to #1397 - which also should be included. The idea is that they are inserted, but keep their scope (which needs to be clarified further).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 24 Jun 2014 11:16 UTC**
These tickets are important, but cannot be resolved for 3.3r1.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 4 Sep 2014 11:36 UTC**
Replying to [comment:17 hansolsson]:
Has there been any progress on this issue? I realize that it probably won't be resolved until at least the next design meeting, but it would nonetheless be good to have a proposal that could be discussed before the meeting. This is a complex issue that I find very important to resolve asap, and if there's no clear idea on how to fix it before the next meeting I fear it will be delayed even further.

Anyway, a question about this:
> Turning the above into spec text:
> 
> 5.6.1 Partial Flattening
> First the names of declared local classes and components are found. Here modifiers are merged to the local elements. Then base-classes are looked up, partially flattened (in the same way) and inserted into the class.
> 
> After the partial flattening it is possible to perform lookup in the class, since the name of all local and inherited elements are known. 

Since lookup in the class is not possible until partial flattening is done, does that mean that classes inside local classes can't be extended from? I.e.:
```
model A
  extends B.C; // B.C can't be found until A has been partially flattened?

  model B
    model C
    end C;
  end B;
end A; 
```

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Sep 2014 16:35 UTC**
Replying to [comment:22 perost]:
> Replying to [comment:17 hansolsson]:
> Has there been any progress on this issue? I realize that it probably won't be resolved until at least the next design meeting, but it would nonetheless be good to have a proposal that could be discussed before the meeting. This is a complex issue that I find very important to resolve asap, and if there's no clear idea on how to fix it before the next meeting I fear it will be delayed even further.
> 
> Anyway, a question about this:
> > Turning the above into spec text:
> > 
> > 5.6.1 Partial Flattening
> > First the names of declared local classes and components are found. Here modifiers are merged to the local elements. Then base-classes are looked up, partially flattened (in the same way) and inserted into the class.
> > 
> > After the partial flattening it is possible to perform lookup in the class, since the name of all local and inherited elements are known. 
> Since lookup in the class is not possible until partial flattening is done, does that mean that classes inside local classes can't be extended from? I.e.:

It means that the procedure have to be described in more detail, it is described as being a two-part procedure ("First" and "Then") but they need to be separated more.

So the dead-lock above is avoided since the lookup of base-classes in 5.6.1 ignores inherited classes (good since inheritance is not yet done). Afterwards base-classes are looked up again - and it is an error if this lookup gives a different result (etc as previously discussed).

Basically 5.6.1 should be split into these two steps: 
* First finding local names (and merging modifiers).
* Then adding base-class contents.
And the first step should have a name.

Slightly updated example:
```
 model A
   extends B.C; // B.C can't be found until A has been partially flattened?
 
   model B
     model C
       extends D; // To make it fun!
     end C;
   end B;
 end A; 
 model D "Nothing to see here"
   ...
 end D;
```
When partially flattening A:
1. We find that A contains a model B and an extends class.
2. Normal lookup of B.C is comprised of finding B in A or its enclosing scopes, and then a C inside this B.
3. Lookup of B in A finds the local class B.
4. We now have to flatten B (will have to verify exactly how much is needed) to find if it contains a C.

after a bit we then begin to partially flatten C:
6. We find that C contains an extends class.
7. We try to find D in A.B.C and go out the enclosing scopes until we find D (at the top-level), and inserts its contents to C.

8. Now we can inserts the contents of (B.)C to A.
9. And finally the partial flattening of A can be completed, and lookup of B.C will still give the same result as before (unless there is a local "B" in "D" - which would be an error; similarly if there is a D in D it would also be an error - but in C).

Thus the specification text would be something like:
5.6.1.1: First the names of declared local classes and components are found. Here modifiers are merged to the local elements. 
5.6.1.2. Then base-classes are looked up, (5.6.1 is applied to them) and inserted into the class. The lookup of the base-classes should be…
5.6.2. All elements are flattened, starting by partially flattening them.
As part of this classes for component elements are looked up, i.e. modifiers including redeclarations for component take effect and the flattened components are inserted into the resulting class.

After step 5.6.1 it is possible to perform lookup in the class, since all named elements are known. It is sometimes necessary to performing lookup before step 5.6.1.2 is complete to handle base-classes, but the result needs to be verified after 5.6.1.2 is complete.

(Step 5.6.2 should also be reformulated, so that there is one step that modifiers take effect and a final step that will ensure that all elements are flattened (some may already have been flattened).)

Most OO-languages avoid this by having the base-classes outside the class itself, which normally prevents inheritance from local class and require a special syntax for separating base-classes from the rest of the contents, and changes to the base-classes are done by overriding virtual functions instead of using modifiers on the extends-clauses. (Import-statements also complicate this, if we want the import-statement to be class-specific, and still be used for finding base-classes.)

**Comment by modelica-trac-importer** [comment]
**Modified by perost on 10 Dec 2014 14:46 UTC**

**Comment by HansOlsson** [comment]
This should by clarified by the new flattening #1829 - if not it should be re-opened with a reference to the new specification.
